apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabaseBase.cs

```diff
@@ -0,0 +1,177 @@
+    /// </summary>
+    public virtual string SpannerPort { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_PORT", null);
+
+    /// <summary>
+    /// The Spanner Instance name to connect to. It is read from the environment variable "TEST_SPANNER_INSTANCE".
+    /// </summary>
+    public virtual string SpannerInstance { get; } = GetEnvironmentVariableOrDefault("TEST_SPANNER_INSTANCE", "spannerintegration");
+
+    /// <summary>
+    /// The Spanner Database name to connect to.
+    /// </summary>
+    public abstract string SpannerDatabase { get; }
+
+    /// <summary>
+    /// The generated name for the database.
+    /// The derived class must provide its implementation to return the database name.
+    /// </summary>
+    public abstract string GeneratedDatabaseName { get; }
+
+    // This is the simplest way of checking whether the environment variable was specified or not.
+    // It's a little ugly, but simpler than the alternatives.
+    /// <summary>
+    /// Returns true if the database was created just for this test, or false if the database was an existing one
+    /// specified through an environment variable.
+    /// </summary>
+    public virtual bool Fresh => SpannerDatabase == GeneratedDatabaseName;
+
+    /// <summary>
+    /// Connection string including database.
+    /// </summary>
+    public string ConnectionString { get; }
+
+    /// <summary>
+    /// Connection string without the database.
+    /// </summary>
+    public string NoDbConnectionString { get; }
+
+    /// <summary>
+    /// The GCP Project ID to be used for the test database.
+    /// </summary>
+    public string ProjectId { get; }
+
+    /// <summary>
+    /// The fully-qualified database name.
+    /// </summary>
+    public DatabaseName DatabaseName { get; }
+
+    internal SpannerClientCreationOptions SpannerClientCreationOptions { get; }
+
+    protected SpannerTestDatabaseBase(string projectId, EmulatorDetection emulatorDetection)
+    {
+        TestLogger.Install();
+        ProjectId = projectId;
+        var builder = new SpannerConnectionStringBuilder
+        {
+            Host = SpannerHost,
+            DataSource = $"projects/{ProjectId}/instances/{SpannerInstance}",
+            EmulatorDetection = emulatorDetection
+        };
+        if (SpannerPort != null)
+        {
+            builder.Port = int.Parse(SpannerPort);
+        }
+        NoDbConnectionString = builder.ConnectionString;
+        SpannerClientCreationOptions = new SpannerClientCreationOptions(builder);
+        var databaseBuilder = builder.WithDatabase(SpannerDatabase);
+        ConnectionString = databaseBuilder.ConnectionString;
+        DatabaseName = databaseBuilder.DatabaseName;
+        // We create database instance only on the emulator.
+        // In production, we expect the instance to already exist.
+        MaybeCreateInstanceOnEmulator(projectId);
+
+        if (Fresh)
```

Comment:
I think this should be, it has probably always worked because where we run with the emulator the environment variable with the DB name is not set, but still, this is a bug.
```suggestion
        if (Fresh || SpannerClientCreationOptions.UsesEmulator)
```

Reply:
Ok. Thanks for letting me know.

Reply:
Not sure - if we use the emulator but set up the table in a different tool for some reason, we wouldn't want to recreate it, would we?

Reply:
@Rishabh-V, @jskeet and I talked about this offline and it's slightly more complicated. So that we don't mix fixing this in this PR, continue asking for Fresh only here, and please create an issue so that we remember to look at this separately. This is test code so its not a huge problem, and also, this bug wouldn't manifest itself in any of our current CI environments.

Reply:
Sure. Thanks Amanda. I will log an issue.